shark_turbine/kernel/wave/expansion.py

```diff
@@ -208,14 +214,40 @@ def expand_graph(
                 dim: val for dim, val in zip(dim_scaling.keys(), dim_combination)
             }
             logger.debug(f"Starting expansion at leaf:{node} in dims:{expand_dims}")
-            _expand_node(node, trace, expand_dims, dim_scaling, expansion_context)
+            _expand_node(
+                node, trace, expand_dims, dim_scaling, dim_tile_size, expansion_context
+            )
+
+    # Replace symbolic tile sizes with actual tile sizes.
+    def replace_index(node: fx.Node) -> bool:
+        if node.index is not None:
+            for dim in node.index:
+                node.index[dim].start = node.index[dim].start.subs(tile_size_mapping)
```

Comment:
Any thoughts on setting the key-value pair for `tile_size_mapping` into indexing_context? Then we can just do all the substitution in one go during codegen. :)

Reply:
Good question. So let me explain what is going on here. So during expansion, we expand a node or unroll it. So for example if we had a read with shape [M, N] and M and N each had a dim_scaling of 2, we would produce read_0_0, read_0_1, read_1_0 and read_1_1 (not necessarily in that order). Now for indexing we want to multiply the current dimension value with the tile size. So in theory, we could just get the value of the dims, so for example for read_1_0, M = 1, N = 0 and then multiply by tile sizes and we are done. This works but the problem is that we visit the same node multiple dims during the DFS. As a result, if we just did the naive approach, we would end up double counting (or more depending on the structure of the graph). In order to avoid this, we use a symbolic tile size and before setting the index we zero out the symbolic tile size to avoid over-counting and then apply the symbolic tile size times the value of M/N. Then as a final step we replace the symbolic value with the actual values. In general, I don't think we can know how many times a node will be traversed so this seemed like a foolproof approach to me. @martin-luecke might have some thoughts on this as well. But the reason why we dont want to add this into indexing_context is because this is temporary index only required during the expansion phase.